File: countlines.py

# ---- Count Lines of Code in Python, and JavaScript Files ----


def count_lines(file_path):
    '''
    Count Lines of Code in Python, and JavaScript Files
    ------------------
    Pass in __file__ if running from a Python target file,
    otherwise pass in path

    Prints
    ------
    * File Name/Path
    * Lines of Code
    * Lines that are not code
    * Total Lines
    * Blank Lines
    * Lines that are just comments (counts line as code if comment is after code)

    Note: Does not count last line if blank. Counts opening and closing of doc strings,
    and single comment symbols as comments.
    '''

    print('---------------- count_lines --------------------------')
    print('File Name/Path: ', file_path)

    code_lines = 0
    not_code_lines = 0
    only_comments = 0
    blank_lines = 0
    doc_switch = False


    if file_path[-3:] == '.js':
        comm_symbol = '//'
        f = open(file_path, "r")
        lines = f.read().splitlines()

        for i in lines:
            k = i.strip()

            # Handle doc strings as comments
            if k[:2] == '/*' or doc_switch == True:
                not_code_lines += 1
                if k != '':
                    only_comments += 1
                else:
                    blank_lines += 1
                doc_switch = True
                if '*/' in k:
                    doc_switch = False
                continue

            if k != '':
                if k[:2] != comm_symbol:
                    code_lines += 1
                else:
                    not_code_lines += 1
                    only_comments += 1
            else:
                not_code_lines += 1
                blank_lines += 1

        f.close()

    else:
        comm_symbol = '#'
        f = open(file_path, "r")
        lines = f.read().splitlines()

        for i in lines:
            k = i.strip()

            # Handle doc strings as comments
            # Turn off
            if "'''" in k and doc_switch == True or '"""' in k and doc_switch == True:
                doc_switch = False
                not_code_lines += 1
                only_comments += 1
                continue

            # Turn on
            if k[:3] == "'''" and doc_switch == False or k[:3] == '"""' and doc_switch == False:
                if "'''" in k[3:] or '"""' in k[3:]:
                    not_code_lines += 1
                    only_comments += 1
                    continue
                else:
                    doc_switch = True

            if doc_switch == True:
                not_code_lines += 1
                if k != '':
                    only_comments += 1
                else:

                    blank_lines += 1
            else:
                if k != '':
                    if k[0] != comm_symbol:
                        code_lines += 1
                    else:
                        not_code_lines += 1
                        only_comments += 1
                else:
                    not_code_lines += 1
                    blank_lines += 1

        f.close()


    print('Lines of Code: ', code_lines)
    print('Lines that are not code: ', not_code_lines)
    print('Total Lines: ', len(lines))
    print('\nBlank Lines: ', blank_lines)
    print('Lines that are just comments: ', only_comments)
    print('------------------------------------------')

    # print(lines)

if __name__ == "__main__":
    # When using here run this, with path inside
    count_lines('')

File: Countlines.cs

class Countlines
{
    static void Main()
    {
        
        int code_lines = 0;
        int not_code_lines = 0;
        int only_comments = 0;
        int blank_lines = 0;
        bool doc_switch = false;

        string filePath = "C:\\Users\\murph\\Desktop\\cvtools.py";

        string comm_symbol = "#";
        
        string[] lines = File.ReadAllLines(@filePath);       // is equiv to System.IO.File.ReadAllLines..

        foreach (string i in lines)                                   // can use "var" var in c# is auto in c++
        {
            string k = i.Trim();

            if (k.Contains("'''") && doc_switch == true || k.Contains("\"\"\"") && doc_switch == true)
            {
                // Console.WriteLine(k);
                doc_switch = false;
                not_code_lines += 1;
                only_comments += 1;
                continue;
            }
            
            // Turn on
            if (k.Length >= 3)
            {
                if (k.Substring(0, 3) == "'''" && doc_switch == false || k.Substring(0, 3) ==
                    "\"\"\"" && doc_switch == false)
                {
                    if (k.Substring(3).Contains("'''") || k.Substring(3).Contains("\"\"\""))
                    {
                        not_code_lines += 1;
                        only_comments += 1;
                        continue;
                    }
                    else
                    {
                        doc_switch = true;
                    }
                }
            }

            if (doc_switch == true)
            {
                    not_code_lines += 1;
                    
                    if (k != "")
                    {
                        only_comments += 1;
                    }
                    else
                    {
                        blank_lines += 1;
                    }
            }

            else
            {
                if (k != "")
                {
                        
                    if (k.Substring(0, 1) != comm_symbol)
                    { 
                        code_lines += 1;
                    }
                    else
                    { 
                        not_code_lines += 1;
                        only_comments += 1;
                    }
                        
                }
                else
                {
                    not_code_lines += 1;
                    blank_lines += 1;
                }
            }
            
        }
        
        Console.WriteLine("Lines of Code: " + code_lines);
        Console.WriteLine("Lines that are not code: " + not_code_lines);
        Console.WriteLine("Total Lines: " + lines.Length);
        Console.WriteLine("\nBlank Lines: " + blank_lines);
        Console.WriteLine("Lines that are just comments: " + only_comments);

        Console.WriteLine("------------------------------------------");
        
    }
}
